main/src/cgeo/geocaching/ui/dialog/CoordinatesCalculateDialog.java

```diff
@@ -0,0 +1,1101 @@
+            updateResult();
+        }
+
+        @Override
+        public void onNothingSelected(final AdapterView<?> arg0) {
+            // do nothing
+        }
+
+    }
+
+    private class HemisphereClickListener implements View.OnClickListener {
+
+        @Override
+        public void onClick(final View view) {
+            final Button button = (Button) view;
+            final CharSequence text = button.getText();
+            if (StringUtils.isBlank(text)) {
+                button.setText("*");
+                return;
+            }
+            switch (text.charAt(0)) {
+                case 'N':
+                    button.setText("S");
+                    break;
+                case 'S':
+                    button.setText("N");
+                    break;
+                case 'E':
+                    button.setText("W");
+                    break;
+                case 'W':
+                    button.setText("E");
+                    break;
+                default:
+                    if (button.equals(bLatHem)) {
+                        button.setText("N");
+                    } else if (button.equals(bLonHem)) {
+                        button.setText("E");
+                    } else {
+                        button.setText(ERROR_CHAR);
+                    }
+
+                    break;
+            }
+
+            updateResult();
+        }
+    }
+
+    /**
+     *
+     * @param gp: Geopoint representing the coordinates from the CoordinateInputDialog
+     * @param calculationState: State to set the calculator to when created
+     */
+    public static CoordinatesCalculateDialog getInstance(final Geopoint gp, final CalcState calculationState) {
+
+        final Bundle args = new Bundle();
+
+        if (gp != null) {
+            args.putParcelable(GEOPOINT_ARG, gp);
+        }
+
+        final CoordinatesCalculateDialog ccd = new CoordinatesCalculateDialog();
+        ccd.setArguments(args);
+        ccd.setSavedState(calculationState);
+        return ccd;
+    }
+
+    @Override
+    public void onCreate(final Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        equations = new ArrayList<>();
+        freeVariables = new ArrayList<>();
+        gp = getArguments().getParcelable(GEOPOINT_ARG);
+        if (gp == null) {
+            gp = Sensors.getInstance().currentGeo().getCoords();
+        }
+        if (savedInstanceState != null) {
+            if (savedInstanceState.getParcelable(GEOPOINT_ARG) != null) {
+                gp = savedInstanceState.getParcelable(GEOPOINT_ARG);
+            }
+
+            final byte[] bytes = savedInstanceState.getByteArray("calc_state");
+            if (bytes != null) {
+                setSavedState((CalcState) SerializationUtils.deserialize(bytes));
+            }
+        }
+    }
+
+    @Override
+    public View onCreateView(final LayoutInflater inflater, final ViewGroup container, final Bundle savedInstanceState) {

```

Comment:
I'm not sure where the complexity comes from.  It's long but really is very simple.
I could break it up into separate sections but it wouldn't really do anything other than run one section after the other.  There's basically just nothing much going on in there.